src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1640,55 +1751,150 @@ public void ReturnConnection(HttpConnection connection)
-                {
-                    Http2Connection[] newHttp2Connections = new Http2Connection[localHttp2Connections.Length - 1];
+                    isNewConnection = false;
 
-                    if (invalidatedIndex > 0)
+                    if (!_http2RequestQueue.TryDequeueNextRequest(connection))
                     {
-                        Array.Copy(localHttp2Connections, newHttp2Connections, invalidatedIndex);
+                        connection.ReleaseStream();
+                        break;
                     }
 
-                    if (invalidatedIndex < localHttp2Connections.Length - 1)
+                    if (NetEventSource.Log.IsEnabled()) connection.Trace("Dequeued waiting HTTP/2 request.");
+                }
+
+                if (_disposed)
+                {
+                    // If the pool has been disposed of, we want to dispose the connection being returned, as the pool is being deactivated.
+                    // We do this after the above in order to satisfy any requests that were queued before the pool was disposed of.
+                    Debug.Assert(_associatedHttp2ConnectionCount > (_availableHttp2Connections?.Count ?? 0));
+                    _associatedHttp2ConnectionCount--;
+                    poolDisposed = true;
+                }
+                else if (usable)
+                {
+                    if (_availableHttp2Connections is null)
                     {
-                        Array.Copy(localHttp2Connections, invalidatedIndex + 1, newHttp2Connections, invalidatedIndex, newHttp2Connections.Length - invalidatedIndex);
+                        _availableHttp2Connections = new List<Http2Connection>();
                     }
 
-                    _http2Connections = newHttp2Connections;
+                    // Add connection to the pool.
+                    _availableHttp2Connections.Add(connection);
+                    if (NetEventSource.Log.IsEnabled()) connection.Trace("Put HTTP/2 connection in pool.");
+                    return;
                 }
             }
+
+            if (poolDisposed)
+            {
+                if (NetEventSource.Log.IsEnabled()) connection.Trace("Disposing HTTP/2 connection returned to pool. Pool was disposed.");
+                connection.Dispose();
+                return;
+            }
+
+            Debug.Assert(!usable);
+
+            // We need to wait until the connection is usable again.
+            DisableHttp2Connection(connection);
+        }
+
+        /// <summary>
+        /// Disable usage of the specified connection because it cannot handle any more streams at the moment.
+        /// We will register to be notified when it can handle more streams (or becomes permanently unusable).
+        /// </summary>
+        private void DisableHttp2Connection(Http2Connection connection)
+        {
+            if (NetEventSource.Log.IsEnabled()) connection.Trace("");
+
+            Task.Run(async () =>
```

Comment:
It seems `Task.Run()` pattern is frequently used by this new code, but it's not clear to me how potential exceptions are handled in all such invocations. In example, I don't see a wrapping `try {} catch {}` block here.

Reply:
The code here should never throw. I could add a try/catch and assert this.